### ISSUE TITLE
[Bug]: Activity Log shows empty "from to" for user role changes when there is no role

### ISSUE THREAD

**Issue by lschuyler** [Automattic/jetpack#30046]
### Quick summary

If a user has no role, or you are switching them from a role to no role, the Activity Log will show neither role in the Activity Log entry.

For example, changing a user from 'no role' to 'Administrator' generates this message:
<img width="568" alt="Screenshot 2022-11-21 at 11 48 46 AM" src="https://user-images.githubusercontent.com/4307088/203135759-82f69159-cb2b-407c-b11e-8704038a88fe.png">

### Steps to reproduce

1. On a Jetpack-connected site, change the role of a site user from any role to 'no role for this site' and save.
2. Then change the role from 'no role for this site' to any other role and save.
3. Check the Activity Log on WordPress.com. You'll see two empty "from to" messages recorded.
4. Back on the site's dashboard, change the role again, from any role to another (don't use 'no role for this site') and save.
5. Check the Activity Log on WordPress.com again and you'll see the expected role change entry with two roles included in the 'from to'.

### What you expected to happen

If a user has no role, or is moving to no role, I'd expect the Activity Log to handle the lack of a role and say "from no role to Administrator" (or vice versa).

### What actually happened

When a user has no role or is switched from a role to not having a role, the Activity Log displays an empty "from to" message.

### Browser

Google Chrome/Chromium

### Platform (Simple, Atomic, or both?)

Self-hosted

### Other notes

I was investigating a report of these empty "from to" entries when using the "User Role Editor" plugin, but I can replicate it when the plugin isn't activated when no role is set, or the user is switched to no role. I don't think use of the User Role Editor is related.

### Reproducibility

Consistent

### Severity

Most (> 50%)

### Available workarounds?

No but the platform is still usable

**Comment by cuemarie** [comment]
📌 SCRUBBING : RESULT - *Replicated / Could Not Replicate / Uncertain*
- Tested on Simple - Replicated
- Tested on Atomic - Replicated
- Tested on Jetpack Connected Site – Replicated

📌 FINDINGS/SCREENSHOTS/VIDEO
- On my Jetpack connected site, I changed the role of user `Ari` from `Subscriber` to `--No role on this site--`.  When I refreshed my WordPress.com Activity Log, the result showed the log with no user roles, as described.
- I can also replicate on Simple and AT sites by switching from Default to Classic View of the users page, and updating any team members in the same manner.

**Screenshots/Recordings**

![setup](https://user-images.githubusercontent.com/27249804/231303149-c7bf58ee-58f7-4794-8615-5de03c038ca7.png)

![Compared to normal role to role changes](https://user-images.githubusercontent.com/27249804/231303597-21a86427-303e-4e20-80e1-5bf56d8bde57.png)

**Free DotCom site example**

On my free WordPress.com site, the Activity Log also logs the change several times (assuming this is reflecting a WPcom cache level user change as well - but it does make the issue appear worse)

![Markup on 2023-04-11 at 15:50:56](https://user-images.githubusercontent.com/27249804/231305468-a9371cc8-4e92-4b81-93b5-c69a27b1680e.png)


📌 ACTIONS
- Assigned to Other
- @jeherve I'm not sure if this should stay in /wp-calypso or in /jetpack, since the Activity Log is only accessible via Calypso, but is a Jetpack feature.  I've transferred it, but feel free to transfer back of course!

**Comment by mreishus** [comment]
I found more info on this. 

### Three Entries Generated for one role change

In general, a role change now generates three entries. One of them is fully populated, and the other two have blank role names.  Example:

![2024-02-26_16-01](https://github.com/Automattic/jetpack/assets/937354/2a578280-556a-49e4-bf6c-3d83fdbf44d2)

### WP.Org Core Change in 2022

The root cause is [this change](https://github.com/WordPress/wordpress-develop/commit/864ab55b33d) to core during 2022. Before the change, only a `set_user_role` action would be fired on a role change.  Now, three actions are fired: `remove_user_role`, `add_user_role`,  and `set_user_role`.

For example, for that screenshot above, these three happened:
```
class-wp-user.php:672 - do_action("remove_user_role", 1234, "author")
class-wp-user.php:677 - do_action("add_user_role", 1234, "editor")
class-wp-user.php:690 - do_action("set_user_role", 1234, "editor", ["author"])
```

### Possible goal of a fix

I think ideally, they would be logged as "Removed role author from sampleuser", "Added role editor to sampleuser", and (existing) "User role changed sampleuser from author to editor".

### Relevant areas of code

**(1)** Hooks are added [here](https://github.com/Automattic/jetpack/blob/6bee74c7c0519f665d4d6b25c2c1d0e8c7a5270b/projects/packages/sync/src/modules/class-users.php#L130-L132).
**(2)** The flags saved for later processing by the activity log are [here](https://github.com/Automattic/jetpack/blob/6bee74c7c0519f665d4d6b25c2c1d0e8c7a5270b/projects/packages/sync/src/modules/class-users.php#L559-L566).
These are later processed in WPCOM:
**WPCOM Step 3**. fbhepr%2Skers%2Sjcpbz%2Sjc%2Qpbagrag%2Syvo%2Snpgvba%2Qgb%2Qnpgvivgl%2Qybt%2Snpgvivgvrf%2Spynff.npgvivgl%2Qhfre%2Qhcqngrq%2Q%2Qwrgcnpx%2Qflap%2Qfnir%2Qhfre.cuc%3Se%3Q3s282s1n%2383-og
**WPCOM Step 4**.
fbhepr%2Skers%2Sjcpbz%2Sjc%2Qpbagrag%2Syvo%2Snpgvivgl%2Qybt%2Qgb%2Qj3%2Qnpgvivgl%2Qfgernz%2Qvgrz%2Snpgvivgvrf%2Spynff.npgvivgl%2Qfgernz%2Qvgrz%2Q%2Qhfre%2Qebyr%2Qhcqngrq.cuc%3Se%3Q3258sr38-og

### Example run through

I move my user's roles. These three actions fire:
```
class-wp-user.php:672 - do_action("remove_user_role", 1234, "author")
class-wp-user.php:677 - do_action("add_user_role", 1234, "editor")
class-wp-user.php:690 - do_action("set_user_role", 1234, "editor", ["author"])
```

These hooks are triggered:
```
		add_action( 'add_user_role', array( $this, 'save_user_role_handler' ), 10, 2 );
		add_action( 'set_user_role', array( $this, 'save_user_role_handler' ), 10, 3 );
		add_action( 'remove_user_role', array( $this, 'save_user_role_handler' ), 10, 2 );
```

The code is the same in all three:
```
	public function save_user_role_handler( $user_id, $role, $old_roles = null ) {
		$this->add_flags(
			$user_id,
			array(
				'role_changed'  => true,
				'previous_role' => $old_roles,
			)
		);
		do_action( 'jetpack_sync_save_user', $user_id, $this->get_flags( $user_id ) );
```

Only one of the calls provides $old_roles, so we generate these activity log flags for processing:
```
    [action_name] => jetpack_sync_save_user
    [args] => Array
        (
            [0] => 1234
            [1] => Array ( [role_changed] => 1 [previous_role] => null )
        )
```

```
    [action_name] => jetpack_sync_save_user
    [args] => Array
        (
            [0] => 1234
            [1] => Array ( [role_changed] => 1 [previous_role] => null )
        )
```

```
    [action_name] => jetpack_sync_save_user
    [args] => Array
        (
            [0] => 1234
            [1] => Array
                (
                    [role_changed] => 1
                    [previous_role] => Array ( [0] => editor )
                )
        )
```

Those objects are saved, and then eventually rendered by the WordPress.com code looking for the "role_changed" flag. It only finds previous_role information for 1 of the three, so this is rendered:

![2024-02-26_16-01](https://github.com/Automattic/jetpack/assets/937354/579c3785-7a2e-40ff-9bdf-36de90a3fb94)

### Full test on WordPress.com

* Have a simple site with an additional user on it you do not mind changing their role.
* Using the WordPress.com console, request wpcom/v2 `/sites/<siteid>/activity?aggregate=true&number=1000`. Look through current->orderedItems.
* Using `/wp-admin/user-edit.php?user_id=<userid>` on the simple site, or just by using the Users -> All Users -> Pick a user and <Edit> links, Change the user's role and save
* Using the WordPress.com console, request wpcom/v2 `/sites/<siteid>/activity?aggregate=true&number=1000`. Look through current->orderedItems. You'll now see three new items instead of one new item.

### Way forward

I think, changing the logging here could be a good first step:
```
	public function save_user_role_handler( $user_id, $role, $old_roles = null ) {
		$this->add_flags(
			$user_id,
			array(
				'role_changed'  => true,
				'previous_role' => $old_roles,
			)
		);
```

Maybe the `set_user_role` could keep the same format, but `remove_user_role` and `set_user_role` would change it up. Perhaps they could drop the `previous_role` key and add their own keys.  Once that's done, I can work on the WP.com side to render it properly.

**Comment by fgiannar** [comment]
Fixed via D140384-code
